Scripts/Runtime/Jobs/ValueAccessController/AbstractValueAccessController.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using Anvil.CSharp.Core;
+using Unity.Collections;
+using Unity.Entities;
+using Unity.Jobs;
+using UnityEngine;
+
+namespace Anvil.Unity.DOTS.Jobs
+{
+    /// <summary>
+    /// A wrapper class for managing async access to a value.
+    /// </summary>
+    /// <remarks>
+    /// Take care that value type implementations propogate state through copies (usually by wrapping a pointer).
+    /// For Example: <see cref="NativeArray{T}" />
+    /// </remarks>
+    public abstract class AbstractValueAccessController<T> : AbstractAnvilBase
+    {
+        private enum AcquisitionState
+        {
+            Unacquired,
+            ReadOnly,
+            ReadWrite,
+        }
+
+        private readonly T m_Value;
+        /// <summary>
+        /// The handle to wait on before <see cref="m_Value"> can be read.
+        /// </summary>
+        private JobHandle m_ReadAccessDependency = default;
+        /// <summary>
+        /// The handle to wait on before <see cref="m_Value"> can be written to.
+        /// </summary>
+        private JobHandle m_WriteAccessDependency = default;
+        private AcquisitionState m_State = AcquisitionState.Unacquired;
+
+
+        /// <summary>
+        /// Creates a new <see cref="AbstractValueAccessController{T}"/> with a given initial value.
+        /// </summary>
+        internal AbstractValueAccessController(T initialValue)
+        {
+            m_Value = initialValue;
+        }
+
+        protected override void DisposeSelf()
+        {
+            // NOTE: If these asserts trigger we should think about calling Complete() on these job handles.
+            Debug.Assert(m_ReadAccessDependency.IsCompleted, "The read access dependency is not completed");
+            Debug.Assert(m_WriteAccessDependency.IsCompleted, "The write access dependency is not completed");
+            (m_Value as IDisposable)?.Dispose();
```

Comment:
Should we restrict `T` to being `IDisposable`?

Reply:
No because not all value types are disposable.

Reply:
True, but I thought this was for the use of wrapping `NativeXXXX` types? 

Would you ever need a read/write dependency for an `int` for example?

Reply:
You could potentially want an `int` wrapped but you'll have to figure out how to deal with the pass by reference issues in the concrete implementation.

In theory you could also use this with managed reference types which wouldn't necessarily be disposable. You couldn't reference the managed type in your job but that managed type may have the data you're passing into your job.